src/lnd.rs

```diff
@@ -0,0 +1,218 @@
+            result => {
+                result?;
+                Ok(())
+            }
+        }
+    }
+
+    /// Obtains a new bitcoin bech32 address from the our lnd node.
+    pub async fn new_bech32_address(&self) -> Result<Address, LndError> {
+        let mut client = self.0.lock().await;
+        let response =
+            client.new_address(NewAddressRequest { r#type: 0, account: String::new() }).await?;
+        response.get_ref().address.parse::<Address>().map_err(LndError::ParseBitcoinAddressFailed)
+    }
+
+    /// Requests to open a channel with remote node, returning the psbt of the funding transaction.
+    ///
+    /// TODO: This should not panic, have proper error handling.
+    pub async fn open_channel(
+        &self,
+        req: OpenChannelRequest,
+    ) -> Result<Option<PartiallySignedTransaction>, LndError> {
+        let client = &mut *self.0.lock().await;
+        let mut response = client.open_channel(req).await?;
+        let stream = response.get_mut();
+
+        while let Some(OpenStatusUpdate { pending_chan_id, update: Some(update) }) =
+            stream.message().await?
+        {
+            use tonic_lnd::rpc::open_status_update::Update;
+            match update {
+                Update::PsbtFund(ready) => {
+                    // TODO: Do not panic here
+                    let psbt =
+                        PartiallySignedTransaction::consensus_decode(&mut &*ready.psbt).unwrap();
+                    eprintln!(
+                        "PSBT received from LND for pending chan id {:?}: {:#?}",
+                        pending_chan_id, psbt
+                    );
+                    assert_eq!(psbt.unsigned_tx.output.len(), 1);
+
+                    // txouts.extend(psbt.unsigned_tx.output);
+                    return Ok(Some(psbt));
+                }
+                // TODO: do not panic
+                x => panic!("Unexpected update {:?}", x),
+            }
+        }
+        Ok(None)
+    }
+
+    /// Sends the `FundingPsbtVerify` message to remote lnd nodes to finalize channels of given
+    /// channel ids.
+    pub async fn verify_funding<I>(&self, funded_psbt: &[u8], chan_ids: I) -> Result<(), LndError>
+    where
+        I: IntoIterator<Item = ChannelId>,
+    {
+        let handles = chan_ids
+            .into_iter()
+            .map(|chan_id| {
+                let client = self.clone();
+                let req = FundingTransitionMsg {
+                    trigger: Some(Trigger::PsbtVerify(FundingPsbtVerify {
+                        pending_chan_id: chan_id.into(),
+                        funded_psbt: funded_psbt.to_vec(),
+                        skip_finalize: true,
+                    })),
+                };
+                tokio::spawn(async move { client.funding_state_step(req).await })
+            })
+            .collect::<Vec<_>>();
+
+        for handle in handles {
+            handle.await.unwrap()?;
+        }
+
+        Ok(())
+    }
```

Comment:
I've made this logic actually async. We spawn a list of async tasks and collect their handlers into a vector, in which we call `await` on each element.